### ISSUE TITLE
lenght of predict result and labels is differ

### ISSUE THREAD

**Issue by mikeun** [ppwwyyxx/tensorpack#588]
I used callback to evaluate confusion matrix after train epoch.

There is dict to collect labels and predictions.
labels and predictions I get by:

    def get_fetches(self):
        return [self.pred_tensor_name, self.label_tensor_name]

    def _before_train(self):
        self.train_tower = self.trainer.tower_func.towers[0]
        self.outputs = [self.train_tower.get_tensor(name) for name in self.get_fetches()]

    def _trigger_step(self):
        outputs = [tensor.eval() for tensor in self.outputs]
        self.on_fetches(outputs)

    def on_fetches(self, outputs):
        pred, label = outputs
        if pred.shape != label.shape:
            print("pass {}, {} != {}".format(len(self.labels['labels'])+1, pred.shape, label.shape))
        self.labels['labels'].append(label)
        self.labels['predic'].append(pred)

it prints warning if there is different length of training batches - predictions and labels
batch size is 64, so every array of evaluated predictions and labels should be 64 elements long, but I've got:
 33%|###########6                                                         |550/1653[08:30<17:08, 1.07it/s]
pass 551, (64,) != (17,)
 67%|########################6                               |1101/1653[17:01<08:31, 1.08it/s]
pass 1102, (64,) != (17,)
100%|#######################################9|1652/1653[25:33<00:00, 1.08it/s]
pass 1653, (64,) != (17,)
100%|########################################|1653/1653[25:41<00:00, 0.24it/s]

len of labels is 1653
quantity of labels is 105651
len of predic is 1653
quantity of predic is 105792

why do you think eval() method of tensor returns differ quantity of elements one run of 551 (these are 33%, 66% and 99% of epoch)?

**Comment by ppwwyyxx** [comment]
You may have used `BatchData(remainder=False)` in evaluation?

**Comment by mikeun** [comment]
I use BatchData(remainder=True). 
This error was eliminated when I setup BatchData(remainder=False).

one another question:
- if I use eval() method - does it force new batch of data from input queue?

**Comment by ppwwyyxx** [comment]
I don't know what is the "eval" method and what's in there.

And what do you mean by "new batch of data"? What data would be an old batch of data?

**Comment by mikeun** [comment]
I meant tensor.eval() method - just call to get an value of tensor.

program logic is -
- it runs train step of current batch. 
- collect step: I'd like to collect labels and predictions of current batch to save them into corresponded dictionary with lists inside. So I use tensor_of_prediction.eval() and tensor_of_labels.eval() method to get values of labels and predictions. This .eval() method should use current data batch to do such step properly.
the question is - does .eval() method use current input data batch to calculate forward run or it reads next batch of data?

**Comment by ppwwyyxx** [comment]
I see. It uses a new batch of data because it's a new sess.run call. 
What you want is to evaluate the tensor **together with** the training and you'll need to use the `before/after_run` method in callbacks.

**Comment by mikeun** [comment]
Now I use     
def _trigger_step(self):
        outputs = [tensor.eval() for tensor in self.outputs]
        self.on_fetches(outputs)


where is self.outputs - list of tensors. How can I get values of this tensors in after_run method?

**Comment by ppwwyyxx** [comment]
`before/after_run` are just `tf.train.SessionRunHooks`. Refer to TF documentation: https://www.tensorflow.org/api_docs/python/tf/train/SessionRunHook

**Comment by mikeun** [comment]
I've got it works 

here is callback:


import numpy as np
import six

from tensorpack.callbacks.base import Callback
from tensorpack.utils import logger
from tensorpack.utils.stats import ConfusionMatrix
from tensorpack.tfutils.common import get_op_tensor_name
import tensorflow as tf

__all__ = ['AfterTrainEpoch','ConfusionMatrixAfterEpoch']


class AfterTrainEpoch(Callback):

    def _setup_graph(self):
        train_tower = self.trainer.tower_func.towers[0]
        self.outputs = [train_tower.get_tensor(name) for name in self.get_fetches()]

    def _before_train(self):
        pass

    def _before_epoch(self):
        self._before_train_epoch()

    def _before_train_epoch(self):
        pass

    def _before_run(self, _):
        return tf.train.SessionRunArgs(fetches=self.outputs)

    def _after_run(self, _, run_values):
        self.on_fetches(run_values.results)

    def on_fetches(self, outputs):
        self._on_fetches(outputs)

    def _on_fetches(self, outputs):
        raise NotImplementedError()

    def _trigger_epoch(self):
        ret = self._after_train_epoch()
        if ret is None:
            return
        for k, v in six.iteritems(ret):
            try:
                v = float(v)
            except ValueError:
                logger.warn("{} returns a non-scalar statistics!".format(type(self).__name__))
                continue
            else:
                self.trainer.monitors.put_scalar(k, v)

    def _after_train_epoch(self):
        pass

    def get_fetches(self):
        try:
            ret = self._get_fetches()
        except NotImplementedError:
            logger.warn("Inferencer._get_output_tensors was deprecated and renamed to _get_fetches")
            ret = self._get_output_tensors()

        return [get_op_tensor_name(n)[1] for n in ret]

    def _get_fetches(self):
        raise NotImplementedError()

    def _get_output_tensors(self):
        pass


class ConfusionMatrixAfterEpoch(AfterTrainEpoch):
    """
    Compute precision / recall of classification, given the
    prediction vector and the label vector.
    """

    def __init__(self, pred_tensor_name, label_tensor_name, labels_list):
        """
        Args:
            pred_tensor_name(str): name of the 0/1 prediction tensor.
            label_tensor_name(str): name of the 0/1 label tensor.
        """
        self.label_list = labels_list
        self.pred_tensor_name = pred_tensor_name
        self.label_tensor_name = label_tensor_name
        self.stat = ConfusionMatrix()
        self.labels = {
                'labels': [],
                'predic': []
                }

    def _before_train_epoch(self):
        self.labels['labels']=[]
        self.labels['predic']=[]

    def _get_fetches(self):
        return [self.pred_tensor_name, self.label_tensor_name]

    def _on_fetches(self, outputs):
        pred, label = outputs
        self.labels['labels'].append(label)
        self.labels['predic'].append(pred)

    def _after_train_epoch(self):
        self.stat.feed(np.concatenate(self.labels['labels'], axis=0),
                       np.concatenate(self.labels['predic'], axis=0), self.label_list)
        out_dict = {}
        for i in range(self.stat.get_conf_matr.shape[0]):
            sum_recall =  self.stat.get_conf_matr[i,:].sum()
            if not bool(sum_recall):
                recall = -1
            else:
                recall = round(self.stat.get_conf_matr[i,i] / sum_recall * 100, 2)
            sum_precision = self.stat.get_conf_matr[:,i].sum()
            if not bool(sum_precision):
                precision = -1
            else:
                precision = round(self.stat.get_conf_matr[i,i] / sum_precision * 100, 2)
            tab='|'
            for j in range(self.stat.get_conf_matr.shape[1]):
                tab += '{: =6.0f}|'.format(self.stat.get_conf_matr[i,j])
            name = 'train conf matrix of class {}: {}, Recall:{: =5.2f}, Precision'.format(i,tab, recall)
            out_dict[name] = precision
        return out_dict


and conf matrix function I added to your utils.stats

from sklearn.metrics import confusion_matrix
class ConfusionMatrix(object):
    """
    Statistics for one_hot mutliclass binary decision,
    including precision, recall, false positive, false negative
    """

    def __init__(self):
        self.reset()

    def reset(self):
        self.conf_matr = None  # positive label

    def feed(self, pred, label, labels):
        """
        Args:
            pred (np.ndarray): binary array.
            label (np.ndarray): binary array of the same size.
        """
        assert pred.shape == label.shape, "{} != {}".format(pred.shape, label.shape)
        self.conf_matr = confusion_matrix(label, pred, labels)

    @property
    def get_conf_matr(self):
        return self.conf_matr
